### PR TITLE
feat(lobby): add countdown timer that auto-starts when all players ready

### LINKED ISSUE #14: Add game start countdown timer
Depends on #11 and #13 

When all players in the lobby Ready up (Including the lobby owner), a countdown timer should begin. Any player should be able to cancel the countdown timer by becoming un ready, and the game should only begin when the countdown timer finishes.

A player leaving should cancel the countdown timer.

New players should be unable to join after the countdown timer has begun

The lobby owner should still be able to start the game immediately

### PR DESCRIPTION
**Pull request by rmartz**
When all players (including the owner) ready up, a 5-second server-side countdown begins. The game starts automatically when it expires. Any player can cancel the countdown by becoming unready, and the owner can still bypass it with the Start Game button at any time.

## Changes

- **Lobby type/schema**: `countdownStartedAt` field on `Lobby`, `PublicLobby`, and `FirebaseLobbyPublic`. Set via `ServerValue.TIMESTAMP` (server-side clock) when all players are ready; cleared on unready, player leave, or game start.
- **Ready API**: Owner can now toggle ready (restriction removed). The countdown fires only when all players including the owner are ready.
- **Join guard**: `validatePlayerJoin` rejects new players while a countdown is active.
- **Real-time propagation**: `lobbySocket.ts` owner-side cache dedup now includes `countdownStartedAt`.
- **Player list UI**: Live countdown ticker shown to all players when active. Ready button now shown for the owner too.
- **Auto-start**: Owner's browser fires `createGame` when the countdown expires.

Closes #14

---
*Created by Claude Sonnet 4.6*

### CODE REVIEW

**Review by rmartz — COMMENTED** [review]
## Requested changes

- **`.then()` chain in auto-start `setTimeout` callback** (`src/app/[gameMode]/lobby/[lobbyId]/page.tsx` line 155): `void flushConfigSyncRef.current().then(() => { ... })` uses a `.then()` chain, which is prohibited by AGENTS.md ("Use `async/await`, not `.then()` chains"). The fix is to make the `setTimeout` callback an `async` arrow function: `setTimeout(async () => { if (!startGameMutationRef.current.isPending) { await flushConfigSyncRef.current(); startGameMutationRef.current.mutate({ gameMode: actualGameMode }); } }, delay)`. No IIFE is introduced — it's just an async callback.

- **Misleading `allPlayersReady` copy shown to owner** (`src/components/lobby/PlayerList.tsx` lines 269–275): When `countdownStartedAt` is absent and `allNonOwnerPlayersReady` is true, the owner sees `PLAYER_LIST_COPY.allPlayersReady` ("All players ready!"). But in this state the owner has not yet readied up — the owner is the one remaining action needed to trigger the countdown. The copy is factually wrong and misses an opportunity to prompt the owner. The string should convey that the owner needs to ready up, something like `"All players ready — click Ready to start the countdown!"`. Update the copy key value and rename it to something that reflects its new meaning (the existing `waitingForHost` key for non-owners already conveys the right idea for that case).

- **Missing `PlayerList.spec.tsx`** (`src/components/lobby/PlayerList.tsx`): AGENTS.md requires adding or updating a co-located component test when a UI component is modified. `PlayerList` was significantly changed (new countdown display, owner ready-up button, `countdownDurationSeconds` prop) but no `PlayerList.spec.tsx` exists and none was added. Add tests covering at minimum: countdown message shown when `countdownStartedAt` is set; `allPlayersReady`/`waitingForHost` message shown when all non-owner players are ready (no countdown); and ready button rendered for the owner.

- **`CountdownActive` story has an impossible fixture** (`src/components/lobby/PlayerList.stories.tsx` lines 89–107): `readyPlayerIds: ["p1", "p2", "p3", "p4"]` with a 5-player lobby (p1–p5) is a state that cannot occur in production — the server only sets `countdownStartedAt` when _all_ players are ready. The story should set `readyPlayerIds` to all five player IDs (`["p1", "p2", "p3", "p4", "p5"]`) to accurately reflect the condition under which the countdown appears.

## Verdict

Four issues to address before merging: a `.then()`-chain convention violation, misleading owner copy, missing component test, and an impossible story fixture.

---
*Claude Sonnet 4.6*

File: src/components/lobby/PlayerList.tsx [hunk 1]
```diff
@@ -69,10 +71,40 @@ export function PlayerList({
...
+      countdownDurationSeconds -
+        Math.floor((Date.now() - countdownStartedAt) / 1000),
+    );
+  });
+
+  useEffect(() => {
+    if (!countdownStartedAt) {
+      setSecondsLeft(0);
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/components/lobby/PlayerList.tsx:92]: Dismissing: `countdownStartedAt` is set via `ServerValue.TIMESTAMP` (Firebase server time), which is always a positive integer — the value `0` is impossible in practice. The `!countdownStartedAt` falsy check is technically inconsistent with the `!== undefined` render check, but causes no real bug. The inconsistency is not severe enough to require a fix.\n\n— *Claude Sonnet 4.6*

File: src/services/lobby.ts [hunk 2]
```diff
@@ -211,16 +212,26 @@ export async function toggleReady(
...
-    .set(updated.length > 0 ? updated : null);
+  const allPlayerIds = Object.keys(data.public.players ?? {});
+  const allReady =
+    allPlayerIds.length >= 2 &&
+    allPlayerIds.every((id) => updated.includes(id));
+
+  await lobbyRef(lobbyId).update({
+    "public/readyPlayerIds": updated.length > 0 ? updated : null,
+    "public/countdownStartedAt": allReady ? ServerValue.TIMESTAMP : null,
+  });
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/services/lobby.ts:231]: Deferring: the read-modify-write race in `toggleReady` is a pre-existing pattern in this codebase, and adding a Firebase transaction would be a significant scope change. In practice, concurrent toggles during the ready-up phase are rare and any inconsistency self-corrects when the next player toggles. Acceptable for MVP.\n\n— *Claude Sonnet 4.6*

File: src/app/[gameMode]/lobby/[lobbyId]/page.tsx [hunk 3]
```diff
@@ -136,6 +138,29 @@ export default function LobbyPage() {
...
+    const startGame = async () => {
+      if (startGameMutationRef.current.isPending) return;
+      await flushConfigSyncRef.current();
+      startGameMutationRef.current.mutate({ gameMode: actualGameMode });
+    };
+    const id = setTimeout(() => void startGame(), delay);
+    return () => {
+      clearTimeout(id);
+    };
+  }, [isOwner, countdownStartedAt, actualGameMode]);
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/app/[gameMode]/lobby/[lobbyId]/page.tsx:162]: Deferring: moving the auto-start trigger to a server-side job is a larger architectural change out of scope for this PR. The client-side `setTimeout` is an acceptable MVP approach. Clock skew is minimal (owner and server are unlikely to differ by more than a few seconds), and the manual Start Game button provides an escape hatch if the auto-start fails.\n\n— *Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **`.then()` chain in auto-start `setTimeout` callback** — converted to `async`/`await` in commit e2b135b9.
- ✅ **Misleading `allPlayersReady` copy shown to owner** — updated to `"All players ready — click Ready to start the countdown!"` in commit e2b135b9.
- ✅ **Missing `PlayerList.spec.tsx`** — added with all required test cases (countdown active, `allPlayersReady`/`waitingForHost` messages, ready button for owner) in commit e2b135b9.
- ✅ **`CountdownActive` story has an impossible fixture** — `readyPlayerIds` now includes all five player IDs (`["p1", "p2", "p3", "p4", "p5"]`) in commit e2b135b9.

## Copilot inline threads

Three of the four open Copilot threads have been replied to and should be resolved:

- ✅ [PlayerList.tsx line 92 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3173366549) — **Dismissed.** `countdownStartedAt` is always a positive Firebase server timestamp; `0` is impossible, so the `!countdownStartedAt` falsy check causes no real bug. Replied to thread.
- ✅ [services/lobby.ts line 223 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3173366579) — **Deferred.** The read-modify-write race in `toggleReady` is a pre-existing codebase pattern; adding a Firebase transaction is out of scope for this PR. Replied to thread.
- ✅ [page.tsx line 162 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3173366616) — **Deferred.** Moving auto-start to a server-side job is an architectural change out of scope for MVP. The manual Start Game button is an escape hatch. Replied to thread.
- ❌ [services/lobby.ts line 127 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3173366594) — **Still open.** See Requested changes below.

## Requested changes

- **`removePlayer` leaves stale IDs in `readyPlayerIds`** (`src/services/lobby.ts` lines 122–127): When a player leaves, `countdownStartedAt` is correctly cleared, but the departing player's ID remains in `public/readyPlayerIds` indefinitely. If the lobby later fills up again, those phantom IDs count toward the "all players ready" threshold in `toggleReady`, which could trigger a countdown when fewer players are actually ready than intended. The fix is to also write `"public/readyPlayerIds": updatedReadyIds` (filtering out `playerId`) in the same `update()` call in `removePlayer`, mirroring how `playerOrder` is cleaned up there.

## Verdict

One prior item (all four from my original review) and three Copilot threads are resolved. One new correctness issue remains: `removePlayer` must also clean up the leaving player's entry from `readyPlayerIds`.

## Deferred

- `toggleReady` read-modify-write race (concurrent toggles can clobber each other) — pre-existing pattern, deferred to a future hardening pass.
- Client-side `setTimeout` auto-start is susceptible to owner tab being suspended/offline — acceptable for MVP with the manual Start Game button as a fallback.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **`.then()` chain in auto-start `setTimeout` callback** — fixed in e2b135b9; `setTimeout` callback is now an `async` arrow function using `await`.
- ✅ **Misleading `allPlayersReady` copy shown to owner** — updated in e2b135b9 to `"All players ready — click Ready to start the countdown!"`.
- ✅ **Missing `PlayerList.spec.tsx`** — added in e2b135b9 with all required test cases.
- ✅ **`CountdownActive` story has an impossible fixture** — fixed in e2b135b9; `readyPlayerIds` now includes all five player IDs.
- ✅ [src/services/lobby.ts line 132 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3173366594) — fixed in 8d766473; `removePlayer` now also writes `"public/readyPlayerIds"` (filtering out the departing player) in the same `update()` call, and updates the in-memory snapshot to match. Thread resolved.

## Verdict

All prior items resolved and no new issues found in either the incremental or full diff — ready to merge.

> Note: branch is 1 commit behind main (non-conflicting). The coordinator will update it before merging.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

All four items from the [2026-04-30 review](https://github.com/rmartz/hidden-role-game/pull/542#pullrequestreview-4208532549) were resolved in commit e2b135b9, and the `removePlayer`/`readyPlayerIds` cleanup from the [2026-05-01 review](https://github.com/rmartz/hidden-role-game/pull/542#pullrequestreview-4211494027) was resolved in commit 8d766473. Those are all ✅.

A subsequent Copilot pass opened three new inline threads. Two remain open:

- ✅ [src/services/lobby.ts line 233 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3174617044) — **Deferred.** Adding `toggleReady`/`removePlayer` service-layer unit tests is valuable, but the Copilot suggestion conflates the issue with an unrelated race-condition refactor. See Requested changes below for a narrower, focused version of this ask.
- ❌ [src/services/lobby.ts line 144 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3174616977) — **Still open.** `removePlayer` does not clear `data.public.countdownStartedAt` in the in-memory snapshot. See Requested changes.
- ❌ [src/services/lobby.ts line 234 (copilot)](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r3174617008) — **Still open.** `toggleReady` sets `data.public.countdownStartedAt = Date.now()` while Firebase stores `ServerValue.TIMESTAMP`. See Requested changes.

## Requested changes

- **`removePlayer` does not clear `countdownStartedAt` from the in-memory snapshot** (`src/services/lobby.ts` line 144): The Firebase `update()` correctly writes `"public/countdownStartedAt": null`, but the in-memory `data.public` is never updated to match — `data.public.countdownStartedAt` retains whatever value it had before the call. When `firebaseToLobby` is then called on the stale `data.public`, the API response incorrectly reports an active countdown even though it was just cancelled. Add `data.public.countdownStartedAt = undefined;` immediately after the `data.public.readyPlayerIds` update on line 143, mirroring the pattern used for `playerOrder` and `readyPlayerIds`.

- **`toggleReady` in-memory timestamp disagrees with Firebase-stored value** (`src/services/lobby.ts` line 233): `data.public.countdownStartedAt = allReady ? Date.now() : undefined` uses the client clock, but Firebase stores `ServerValue.TIMESTAMP` (the server clock). The immediate API response can therefore report a `countdownStartedAt` that differs from the authoritative value in RTDB, causing the initial countdown display to jump or skip when the subscription catches up. Set `data.public.countdownStartedAt = undefined` when `allReady` is true (mirroring the "don't guess the server value" approach) and let the RTDB subscription propagate the real timestamp to all clients, rather than pre-populating a potentially skewed local value.

- **No service-layer tests for `toggleReady` and `removePlayer` countdown behavior** (`src/services/lobby.ts`): AGENTS.md requires smoke tests for service orchestration. The PR adds a `server/lobby.spec.ts` test for `validatePlayerJoin`, but `src/services/lobby.spec.ts` has no coverage for the new countdown logic in `toggleReady` (sets `countdownStartedAt` when all players ready, clears it on unready) or `removePlayer` (clears `countdownStartedAt` when a player leaves during a countdown). Add at least three tests: countdown started when all players ready, countdown cleared on unready, and countdown cleared when a player removes themselves during a countdown.

## Verdict

Three new issues from the latest Copilot review pass require attention: a correctness bug in `removePlayer` (stale in-memory snapshot), a timestamp skew in `toggleReady`, and missing service-layer tests for the countdown logic.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ [`removePlayer` doesn't clear `countdownStartedAt` in the returned Lobby](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r2074827438) — fixed: clear `data.public.countdownStartedAt` on the in-memory snapshot after writing null to Firebase
- ✅ [`toggleReady` uses `Date.now()` for `countdownStartedAt` instead of deferring to the server timestamp](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r2074827449) — fixed: avoid propagating `countdownStartedAt` in the immediate response; clients receive the authoritative `ServerValue.TIMESTAMP` value via the RTDB subscription
- ✅ [Missing tests for countdown start/clear behavior](https://github.com/rmartz/hidden-role-game/pull/542#discussion_r2074827457) — added three tests covering: all players ready up starts countdown, unreadying clears countdown, and a player leaving during countdown clears it

## Verdict

All three prior items resolved and tests passing — ready to merge.

---
*Claude Sonnet 4.6*

**Comment by rmartz** [comment]
**Pre-merge check**

- ✅ PR ready
- ✅ Review
- ✅ UAT
- ✅ Branch
- ✅ CI

Merging.

---
*Posted by Claude Sonnet 4.6*
